test/hotreload/myrole_test.lua

```diff
@@ -0,0 +1,399 @@
+        return require("cartridge.roles").reload()
+    ]])
+
+    t.assert_equals(ok, nil)
+    t.assert_covers(err, {
+        class_name = 'RegisterRoleError',
+        err = 'F',
+    })
+
+    local resp = g.srv:graphql({query = [[
+        {cluster { self {state error} }}
+    ]]})
+
+    t.assert_equals(resp.data.cluster.self, {
+        state = "ReloadError",
+        error = "F",
+    })
+
+    reload_myrole(function() return {role_name = 'myrole'} end)
+    g.cluster:wait_until_healthy()
+end
+
+function g.test_roledeps()
+    g.srv.net_box:eval([[
+        package.preload["role-a"] = function() return {} end
+        package.preload["role-b"] = function() return {} end
+    ]])
+
+    reload_myrole(function()
+        return {
+            role_name = 'myrole',
+            dependencies = {'role-a'},
+        }
+    end)
+
+    g.srv.net_box:eval([[
+        local cartridge = require('cartridge')
+        assert(package.loaded['role-a'])
+        assert(package.loaded['role-b'] == nil)
+        assert(cartridge.service_get('role-a'))
+        assert(cartridge.service_get('role-b') == nil)
+    ]])
+
+    reload_myrole(function()
+        return {
+            role_name = 'myrole',
+            dependencies = {'role-b'},
+        }
+    end)
+
+    g.srv.net_box:eval([[
+        local cartridge = require('cartridge')
+        assert(package.loaded['role-a'] == nil)
+        assert(package.loaded['role-b'])
+        assert(cartridge.service_get('role-a') == nil)
+        assert(cartridge.service_get('role-b'))
+    ]])
+end
+
+function g.test_rpc()
+    reload_myrole(function()
+        return {
+            role_name = 'myrole',
+            cheer = function() return 'Hello, Cartridge' end,
+        }
+    end)
+
+    -- New RPC works
+    t.assert_equals(g.srv.net_box:call(
+        'package.loaded.cartridge.rpc_call',
+        {'myrole', 'cheer'}
+    ), 'Hello, Cartridge')
+
+    reload_myrole(function()
+        return {
+            role_name = 'myrole',
+            echo = function(...) return ... end,
+        }
+    end)
+
+    -- Old RPC isn't available anymore
+    local ret, err = g.srv.net_box:call(
+        'package.loaded.cartridge.rpc_call',
+        {'myrole', 'cheer'}
+    )
+    t.assert_equals(ret, nil)
+    t.assert_covers(err, {
+        class_name = 'RemoteCallError',
+        err = 'Role "myrole" has no method "cheer"',
+    })
+
+    -- New RPC works
+    t.assert_equals(g.srv.net_box:call(
+        'package.loaded.cartridge.rpc_call',
+        {'myrole', 'echo', {'2020-11-10'}}
+    ), '2020-11-10')
+end
+
+function g.test_netbox()
+    reload_myrole(function()
+        rawset(_G, 'hang', function()
+            rawset(_G, 'test_ready', true)
+            require('log').info('Weird netbox call')
+            require('fiber').sleep(math.huge)
+            return true
+        end)
+        return {role_name = 'myrole'}
+    end)
+
+    local future = g.srv.net_box:call('hang', nil, {is_async = true})
```

Comment:
Be careful with such fibers. A bug (https://github.com/tarantool/tarantool/issues/4834) was recently fixed.

Reply:
It won't affect the test. That bug was about canceling the fiber on the client. But in this test, the server-side fiber is canceled.